File: cs/days/day03.cs

using System.Text.RegularExpressions;

namespace Shunty.AoC.Days;

// https://adventofcode.com/2024/day/3 - Mull It Over

public class Day03 : AocDaySolver
{
    public int DayNumber => 3;
    public string Title => "Mull It Over";

    private const string patternP1 = @"mul\(\d+,\d+\)";
   	private const string patternP2 = @"mul\(\d+,\d+\)|do\(\)|don\'t\(\)";
    private const string patternInner = @"mul\((\d+),(\d+)\)";

    private readonly Regex reP1 = new(patternP1, RegexOptions.Compiled);
    private readonly Regex reP2 = new(patternP2, RegexOptions.Compiled);
    private readonly Regex reInner = new(patternInner, RegexOptions.Compiled);

    public async Task Solve()
    {
        var input = string.Join("", await File.ReadAllLinesAsync(AocUtils.FindInputFile(DayNumber)));
        var matchesP1 = reP1.Matches(input);

        var part1 = 0L;
        foreach (var m in matchesP1.ToList())
        {
            var mm = reInner.Matches(m.Value).ToList();
            part1 += int.Parse(mm[0].Groups[1].Value) * int.Parse(mm[0].Groups[2].Value);
        }
        this.ShowDayResult(1, part1);

        var ok = true;
        var part2 = 0L;
        var matchesP2 = reP2.Matches(input);
        foreach (var m in matchesP2.ToList())
        {
            if (m.Value == "don't()")
            {
                ok = false;
            }
            else if (m.Value == "do()")
            {
                ok = true;
            }
            else if (ok)
            {
                var mm = reInner.Matches(m.Value).ToList();
                part2 += int.Parse(mm[0].Groups[1].Value) * int.Parse(mm[0].Groups[2].Value);
            }
        }
        this.ShowDayResult(2, part2);
    }
}

File: go/days/day03.go
package days

// https://adventofcode.com/2024/day/3 - Mull It Over

import (
	aoc "aoc2024/aocutils"
	"regexp"
)

func Day03() {
	const DAY = 3
	input, err := aoc.GetDayInputString(DAY)
	aoc.CheckError(err)

	reP1 := regexp.MustCompile(`mul\(\d+,\d+\)`)
	matches := reP1.FindAllString(input, -1)

	part1 := 0
	for _, match := range matches {
		part1 += mul(match)
	}
	aoc.DayHeader(DAY)
	aoc.PrintResult(1, part1)

	reP2 := regexp.MustCompile(`mul\(\d+,\d+\)|do\(\)|don\'t\(\)`)
	matches = reP2.FindAllString(input, -1)

	part2 := 0
	do := true
	for _, match := range matches {
		switch match {
		case "do()":
			do = true
		case "don't()":
			do = false
		default:
			if do {
				part2 += mul(match)
			}
		}
	}
	aoc.PrintResult(2, part2)
}

func mul(mulStr string) int {
	s := mulStr[4 : len(mulStr)-1] // cut off 'mul(' and ')'
	nums, err := aoc.StringToInts(s, ",")
	aoc.CheckError(err)
	return nums[0] * nums[1]
}
